addons/dispatch/0.4.x/dispatch-1.yaml

```diff
@@ -0,0 +1,46 @@
+---
+apiVersion: kubeaddons.mesosphere.io/v1beta1
+kind: ClusterAddon
+metadata:
+  name: dispatch
+  labels:
+    kubeaddons.mesosphere.io/name: dispatch
+  annotations:
+    catalog.kubeaddons.mesosphere.io/addon-revision: "0.4.1-1"
+    appversion.kubeaddons.mesosphere.io/dispatch: "0.4.1"
+    endpoint.kubeaddons.mesosphere.io/dispatch: /dispatch/tekton/
+    docs.kubeaddons.mesosphere.io/dispatch: "https://beta-dispatch.d2iq.com/"
+    appversion.kubeaddons.mesosphere.io/argo-cd: "1.2.0"
+    endpoint.kubeaddons.mesosphere.io/argo-cd: /dispatch/argo-cd
+    docs.kubeaddons.mesosphere.io/argo-cd: "https://argoproj.github.io/argo-cd/user-guide/"
+    values.chart.helm.kubeaddons.mesosphere.io/dispatch: "https://raw.githubusercontent.com/mesosphere/charts/master/stable/dispatch/values.yaml"
+spec:
+  namespace: dispatch
+  kubernetes:
+    minSupportedVersion: v1.15.6
+  requires:
+    - matchLabels:
+        kubeaddons.mesosphere.io/provides: storageclass
```

Comment:
@joejulian are we using this or enforcing this? I see velero doesn't but instead using a label `kubeaddons.mesosphere.io/hack-requires-defaultstorageclass: "true"` and works as long as there is a default storage class.

Reply:
I was able to confirm that with that label and removing the two volume parts dispatch was able to deploy:

```
dispatch:
    - "cert-manager"
    - "prometheus"
    - "prometheusadapter"
    - "opsportal"
    - "dispatch"
```

@joejulian @shaneutt probably should discuss this or get some advice from you, i am not exactly well versed in what approach i should suggest here.

Reply:
So I added this because I ran into an issue where Dispatch deployed before the volume provisioner and the PVCs were never bound, so I'm not sure if I should remove them. I'll do a little more testing today and see if I can reproduce that issue to get more details.

Reply:
If I remove the requirements and deploy in Konvoy, dispatch gets deployed before localvolumeprovisioner and I get this:

```
Name:          dispatch-kubeaddons-minio
Namespace:     dispatch
StorageClass:  
Status:        Pending
Volume:        
Labels:        app=minio
               chart=minio-2.5.14
               heritage=Tiller
               release=dispatch-kubeaddons
Annotations:   <none>
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Mounted By:    dispatch-kubeaddons-minio-79cd5cb5f7-6tgf4
Events:
  Type    Reason         Age                  From                         Message
  ----    ------         ----                 ----                         -------
  Normal  FailedBinding  4s (x21 over 4m49s)  persistentvolume-controller  no persistent volumes available for this claim and no storage class is set
```

I'm thinking maybe my best bet is to update kubeaddons test harness to support customizing the kind cluster?

Reply:
check the annotations i suggested above and retest.:

```
  annotations:
    catalog.kubeaddons.mesosphere.io/addon-revision: "0.4.1-1"
    appversion.kubeaddons.mesosphere.io/dispatch: "0.4.1"
    endpoint.kubeaddons.mesosphere.io/dispatch: /dispatch/tekton/
    docs.kubeaddons.mesosphere.io/dispatch: "https://beta-dispatch.d2iq.com/"
    appversion.kubeaddons.mesosphere.io/argo-cd: "1.2.0"
    endpoint.kubeaddons.mesosphere.io/argo-cd: /dispatch/argo-cd
    docs.kubeaddons.mesosphere.io/argo-cd: "https://argoproj.github.io/argo-cd/user-guide/"
    values.chart.helm.kubeaddons.mesosphere.io/dispatch: "https://raw.githubusercontent.com/mesosphere/charts/master/stable/dispatch/values.yaml"
    # TODO: we're temporarily supporting dependency on an existing default storage class
    # on the cluster, this hack will trigger re-queue on Addons until one exists.
    kubeaddons.mesosphere.io/hack-requires-defaultstorageclass: "true"
```

Reply:
https://github.com/kubernetes/kubernetes/issues/44370#issuecomment-343049129

> If at the time you created the PVC, you did not have a default storage class, then the admission controller will not be able to fill in the name of the default storage class into the PVC object.

So I do need the storage to exist first since I can't reliably assume a storage class.

Reply:
I see @joejulian feels strongly about this :D https://github.com/kubernetes/kubernetes/issues/44370#issuecomment-528435944

Reply:
> So I do need the storage to exist first since I can't reliably assume a storage class.

yep, for localvolumeprovisioner that seems to be the case. remove the requires label, and test with the annotation i mention above.

Reply:
![Screen Shot 2020-01-08 at 6 39 50 PM](https://user-images.githubusercontent.com/20777618/72095849-1aa69d00-32ce-11ea-9247-ca6752c9bee3.png)
I didnt catch the exact moment it completed, but here i show you what it looks like when it works and things are being torn down.

Reply:
Right. That works in kind, but the issue I run into is on konvoy itself. It'll work here, but won't work on konvoy (at least, konvoy-in-docker).

Reply:
It requires dynamic provisioning so i suspect it will work on azure and aws. We cannot yet test on a konvoy cluster per se, so we are trying to get this environment as close as possible. Your call about how to proceed.

Reply:
Here's a novel solution, I guess. I can manually create the PV in the test environment:

```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: dispatch-minio-pv
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  local:
    path: /mnt/disks/dispatch-minio-pv
  persistentVolumeReclaimPolicy: Delete
  storageClassName: localvolumeprovisioner
  volumeMode: Filesystem
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/os
          operator: In
          values:
          - linux
```

This seems like the least intrusive method that still maintains the integrity of the test.

Reply:
will konvoy docker need that also to have dispatch working?

Reply:
No - local volumes on konvoy-in-docker work because we attach docker volumes and mount them into /mnt/disks so that the local volume provisioner can dynamically create them.

We can do this with kubeaddons test harness, but I'll need to do some upstream work in kubeaddons to make it work (allow passing in a custom kind configuration file). I noticed we're using an older version of kubeaddons so I'm not sure how much of a pain it is to upgrade.

Reply:
:boom: I added support for adding volumes to kind and everything seems to work!

Reply:
`kubeaddons.mesosphere.io/provides: storageclass` will ensure a storageclass labeled addon exists, it may not be set as default. The safer option is the  `kubeaddons.mesosphere.io/hack-requires-defaultstorageclass` _label_ as demonstrated here:
https://github.com/mesosphere/kubernetes-base-addons/blob/master/addons/traefik/1.7.x/traefik-1.yaml#L9-L11